hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/JDBCSource.java

```diff
@@ -0,0 +1,233 @@
+      dataFrameReader = dataFrameReader
+          .option(Config.RDBMS_TABLE_PROP, properties.getString(Config.RDBMS_TABLE_NAME));
+
+      if (!properties.containsKey(Config.PASSWORD)) {
+        if (properties.containsKey(Config.PASSWORD_FILE)) {
+          if (!StringUtils.isNullOrEmpty(properties.getString(Config.PASSWORD_FILE))) {
+            LOG.info("Reading password for password file {}", properties.getString(Config.PASSWORD_FILE));
+            FileSystem fileSystem = FileSystem.get(new Configuration());
+            passwordFileStream = fileSystem.open(new Path(properties.getString(Config.PASSWORD_FILE)));
+            byte[] bytes = new byte[passwordFileStream.available()];
+            passwordFileStream.read(bytes);
+            dataFrameReader = dataFrameReader.option(Config.PASSWORD_PROP, new String(bytes));
+          } else {
+            throw new IllegalArgumentException(
+                String.format("%s property cannot be null or empty", Config.PASSWORD_FILE));
+          }
+        } else {
+          throw new IllegalArgumentException(String.format("JDBCSource needs either a %s or %s to connect to RDBMS "
+              + "datasource", Config.PASSWORD_FILE, Config.PASSWORD));
+        }
+      } else if (!StringUtils.isNullOrEmpty(properties.getString(Config.PASSWORD))) {
+        dataFrameReader = dataFrameReader.option(Config.PASSWORD_PROP, properties.getString(Config.PASSWORD));
+      } else {
+        throw new IllegalArgumentException(String.format("%s cannot be null or empty. ", Config.PASSWORD));
+      }
+      if (properties.containsKey(Config.EXTRA_OPTIONS)) {
+        if (!StringUtils.isNullOrEmpty(properties.getString(Config.EXTRA_OPTIONS))) {
+          LOG.info("Setting {}", Config.EXTRA_OPTIONS);
+          String[] options = properties.getString(Config.EXTRA_OPTIONS).split(",");
+          for (String option : options) {
+            if (!StringUtils.isNullOrEmpty(option)) {
+              String[] kv = option.split("=");
+              dataFrameReader = dataFrameReader.option(kv[0], kv[1]);
+              LOG.info("{} = {} has been set for JDBC pull ", kv[0], kv[1]);
+            }
+          }
+        }
+      }
+      if (properties.getBoolean(Config.IS_INCREMENTAL)) {
+        DataSourceUtils.checkRequiredProperties(properties, Arrays.asList(Config.INCREMENTAL_COLUMN));
+      }
+      return dataFrameReader;
+    } catch (Exception e) {
+      throw new HoodieException(e);
+    } finally {
+      IOUtils.closeStream(passwordFileStream);
+    }
+  }
+
+  @Override
+  protected Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+    try {
+      DataSourceUtils.checkRequiredProperties(props, Arrays.asList(Config.URL, Config.DRIVER_CLASS, Config.USER,
+          Config.RDBMS_TABLE_NAME, Config.IS_INCREMENTAL));
+
+      Option<String> beginInstant =
```

Comment:
its not an instant right? can you rework variable names based on the jdbc source

Reply:
Renamed to lastCheckpoint